File: csharp/2048-ai.cs
﻿using System;
using System.Collections.Generic;
using System.Threading;

class Game2048
{
    Random rand = new Random();
    const UInt64 ROW_MASK = 0xFFFF;
    const UInt64 COL_MASK = 0x000F000F000F000F;
    const int TABLESIZE = 65536;
    UInt16[] row_left_table = new UInt16[TABLESIZE];
    UInt16[] row_right_table = new UInt16[TABLESIZE];
    UInt32[] score_table = new UInt32[TABLESIZE];
    double[] score_heur_table = new double[TABLESIZE];
    delegate int get_move_func_t(UInt64 board);
    const int UP = 0;
    const int DOWN = 1;
    const int LEFT = 2;
    const int RIGHT = 3;

    struct trans_table_entry_t
    {
        public int depth;
        public double heuristic;
        public trans_table_entry_t(int depth, double heuristic)
        {
            this.depth = depth;
            this.heuristic = heuristic;
        }
    }
    class eval_state
    {
        public Dictionary<UInt64, trans_table_entry_t> trans_table = new Dictionary<UInt64, trans_table_entry_t>();
        public int maxdepth;
        public int curdepth;
        public int nomoves;
        public int tablehits;
        public int cachehits;
        public int moves_evaled;
        public int depth_limit;
    }

    const double SCORE_LOST_PENALTY = 200000.0f;
    const double SCORE_MONOTONICITY_POWER = 4.0f;
    const double SCORE_MONOTONICITY_WEIGHT = 47.0f;
    const double SCORE_SUM_POWER = 3.5f;
    const double SCORE_SUM_WEIGHT = 11.0f;
    const double SCORE_MERGES_WEIGHT = 700.0f;
    const double SCORE_EMPTY_WEIGHT = 270.0f;
    const double CPROB_THRESH_BASE = 0.0001f;
    const UInt16 CACHE_DEPTH_LIMIT = 15;

    Int32 unif_random(Int32 n)
    {
        return rand.Next(n);
    }

    void clear_screen()
    {
        Console.Clear();
    }

    UInt64 unpack_col(UInt16 row)
    {
        UInt64 tmp = row;
        return (tmp | (tmp << 12) | (tmp << 24) | (tmp << 36)) & COL_MASK;
    }

    UInt16 reverse_row(UInt16 row)
    {
        return (UInt16)((row >> 12) | ((row >> 4) & 0x00F0) | ((row << 4) & 0x0F00) | (row << 12));
    }

    void print_board(UInt64 board)
    {
        Console.WriteLine("-----------------------------");
        for (int i = 0; i < 4; i++)
        {
            for (int j = 0; j < 4; j++)
            {
                int power_val = (int)(board & 0xf);

                if (power_val == 0)
                    Console.Write("|{0,6}", ' ');
                else
                    Console.Write("|{0,6}", 1 << power_val);
                board >>= 4;
            }
            Console.WriteLine("|");
        }
        Console.WriteLine("-----------------------------");
    }

    UInt64 transpose(UInt64 x)
    {
        UInt64 a1 = x & 0xF0F00F0FF0F00F0F;
        UInt64 a2 = x & 0x0000F0F00000F0F0;
        UInt64 a3 = x & 0x0F0F00000F0F0000;
        UInt64 a = a1 | (a2 << 12) | (a3 >> 12);
        UInt64 b1 = a & 0xFF00FF0000FF00FF;
        UInt64 b2 = a & 0x00FF00FF00000000;
        UInt64 b3 = a & 0x00000000FF00FF00;

        return b1 | (b2 >> 24) | (b3 << 24);
    }

    int count_empty(UInt64 x)
    {
        x |= (x >> 2) & 0x3333333333333333;
        x |= (x >> 1);
        x = ~x & 0x1111111111111111;
        x += x >> 32;
        x += x >> 16;
        x += x >> 8;
        x += x >> 4;
        return (int)(x & 0xf);
    }

    void init_tables()
    {
        UInt16 row = 0, result = 0;
        UInt16 rev_row = 0, rev_result = 0;
        UInt16[] line = new UInt16[4];

        do
        {
            int i = 0, j = 0;
            UInt32 score = 0;
            line[0] = (UInt16)(row & 0xf);
            line[1] = (UInt16)((row >> 4) & 0xf);
            line[2] = (UInt16)((row >> 8) & 0xf);
            line[3] = (UInt16)((row >> 12) & 0xf);

            for (i = 0; i < 4; ++i)
            {
                int rank = line[i];

                if (rank >= 2)
                    score += (UInt32)((rank - 1) * (1 << rank));
            }
            score_table[row] = score;

            double sum = 0.0f;
            int empty = 0;
            int merges = 0;
            int prev = 0;
            int counter = 0;

            for (i = 0; i < 4; ++i)
            {
                int rank = line[i];

                sum += Math.Pow(rank, SCORE_SUM_POWER);
                if (rank == 0)
                {
                    empty++;
                }
                else
                {
                    if (prev == rank)
                    {
                        counter++;
                    }
                    else if (counter > 0)
                    {
                        merges += 1 + counter;
                        counter = 0;
                    }
                    prev = rank;
                }
            }
            if (counter > 0)
                merges += 1 + counter;

            double monotonicity_left = 0.0f;
            double monotonicity_right = 0.0f;

            for (i = 1; i < 4; ++i)
            {
                if (line[i - 1] > line[i])
                {
                    monotonicity_left +=
                        Math.Pow(line[i - 1], SCORE_MONOTONICITY_POWER) - Math.Pow(line[i], SCORE_MONOTONICITY_POWER);
                }
                else
                {
                    monotonicity_right +=
                        Math.Pow(line[i], SCORE_MONOTONICITY_POWER) - Math.Pow(line[i - 1], SCORE_MONOTONICITY_POWER);
                }
            }

            score_heur_table[row] = SCORE_LOST_PENALTY + SCORE_EMPTY_WEIGHT * empty + SCORE_MERGES_WEIGHT * merges -
                SCORE_MONOTONICITY_WEIGHT * Math.Min(monotonicity_left, monotonicity_right) - SCORE_SUM_WEIGHT * sum;

            for (i = 0; i < 3; ++i)
            {
                for (j = i + 1; j < 4; ++j)
                {
                    if (line[j] != 0)
                        break;
                }
                if (j == 4)
                    break;

                if (line[i] == 0)
                {
                    line[i] = line[j];
                    line[j] = 0;
                    i--;
                }
                else if (line[i] == line[j])
                {
                    if (line[i] != 0xf)
                        line[i]++;
                    line[j] = 0;
                }
            }
            result = (UInt16)(line[0] | (line[1] << 4) | (line[2] << 8) | (line[3] << 12));

            rev_row = reverse_row(row);
            rev_result = reverse_row(result);
            row_left_table[row] = (UInt16)(row ^ result);
            row_right_table[rev_row] = (UInt16)(rev_row ^ rev_result);
        } while (row++ != 0xFFFF);
    }

    UInt64 execute_move(UInt64 board, int move)
    {
        UInt64 ret = board;

        if (move == UP) {
            board = transpose(board);
            ret ^= unpack_col(row_left_table[board & ROW_MASK]);
            ret ^= unpack_col(row_left_table[(board >> 16) & ROW_MASK]) << 4;
            ret ^= unpack_col(row_left_table[(board >> 32) & ROW_MASK]) << 8;
            ret ^= unpack_col(row_left_table[(board >> 48) & ROW_MASK]) << 12;
        } else if (move == DOWN) {
            board = transpose(board);
            ret ^= unpack_col(row_right_table[board & ROW_MASK]);
            ret ^= unpack_col(row_right_table[(board >> 16) & ROW_MASK]) << 4;
            ret ^= unpack_col(row_right_table[(board >> 32) & ROW_MASK]) << 8;
            ret ^= unpack_col(row_right_table[(board >> 48) & ROW_MASK]) << 12;
        } else if (move == LEFT) {
            ret ^= (UInt64)(row_left_table[board & ROW_MASK]);
            ret ^= (UInt64)(row_left_table[(board >> 16) & ROW_MASK]) << 16;
            ret ^= (UInt64)(row_left_table[(board >> 32) & ROW_MASK]) << 32;
            ret ^= (UInt64)(row_left_table[(board >> 48) & ROW_MASK]) << 48;
        } else if (move == RIGHT) {
            ret ^= (UInt64)(row_right_table[board & ROW_MASK]);
            ret ^= (UInt64)(row_right_table[(board >> 16) & ROW_MASK]) << 16;
            ret ^= (UInt64)(row_right_table[(board >> 32) & ROW_MASK]) << 32;
            ret ^= (UInt64)(row_right_table[(board >> 48) & ROW_MASK]) << 48;
        }
        return ret;
    }

    UInt32 score_helper(UInt64 board)
    {
        return score_table[board & ROW_MASK] + score_table[(board >> 16) & ROW_MASK] +
            score_table[(board >> 32) & ROW_MASK] + score_table[(board >> 48) & ROW_MASK];
    }

    double score_heur_helper(UInt64 board)
    {
        return score_heur_table[board & ROW_MASK] + score_heur_table[(board >> 16) & ROW_MASK] +
            score_heur_table[(board >> 32) & ROW_MASK] + score_heur_table[(board >> 48) & ROW_MASK];
    }

    UInt32 score_board(UInt64 board)
    {
        return score_helper(board);
    }

    double score_heur_board(UInt64 board)
    {
        return score_heur_helper(board) + score_heur_helper(transpose(board));
    }

    UInt64 draw_tile()
    {
        return (UInt64)((unif_random(10) < 9) ? 1 : 2);
    }

    UInt64 insert_tile_rand(UInt64 board, UInt64 tile)
    {
        int index = unif_random(count_empty(board));
        UInt64 tmp = board;

        while (true)
        {
            while ((tmp & 0xf) != 0)
            {
                tmp >>= 4;
                tile <<= 4;
            }
            if (index == 0)
                break;
            --index;
            tmp >>= 4;
            tile <<= 4;
        }
        return board | tile;
    }

    UInt64 initial_board()
    {
        UInt64 board = (UInt64)(draw_tile()) << (unif_random(16) << 2);

        return insert_tile_rand(board, draw_tile());
    }

    int get_depth_limit(UInt64 board)
    {
        UInt16 bitset = 0;
        int max_limit = 3;
        int count = 0;

        while (board != 0)
        {
            bitset |= (UInt16)(1 << (int)(board & 0xf));
            board >>= 4;
        }

        if (bitset <= 2048) {
            return max_limit;
        } else if (bitset <= 2048 + 1024) {
            max_limit = 4;
        } else if (bitset <= 4096) {
            max_limit = 5;
        } else if (bitset <= 4096 + 2048) {
            max_limit = 6;
        } else if (bitset <= 8192) {
            max_limit = 7;
        } else {
            max_limit = 8;
        }

        bitset >>= 1;
        while (bitset != 0)
        {
            bitset &= (UInt16)(bitset - 1);
            count++;
        }
        count -= 2;
        count = Math.Max(count, 3);
        count = Math.Min(count, max_limit);
        return count;
    }

    double score_tilechoose_node(ref eval_state state, UInt64 board, double cprob)
    {
        if (cprob < CPROB_THRESH_BASE || state.curdepth >= state.depth_limit)
        {
            state.maxdepth = Math.Max(state.curdepth, state.maxdepth);
            state.tablehits++;
            return score_heur_board(board);
        }
        if (state.curdepth < CACHE_DEPTH_LIMIT)
        {
            if (state.trans_table.ContainsKey(board))
            {
                trans_table_entry_t entry = state.trans_table[board];

                if (entry.depth <= state.curdepth)
                {
                    state.cachehits++;
                    return entry.heuristic;
                }
            }
        }

        int num_open = count_empty(board);

        cprob /= num_open;

        double res = 0.0f;
        UInt64 tmp = board;
        UInt64 tile_2 = 1;

        while (tile_2 != 0)
        {
            if ((tmp & 0xf) == 0)
            {
                res += score_move_node(ref state, board | tile_2, cprob * 0.9f) * 0.9f;
                res += score_move_node(ref state, board | (tile_2 << 1), cprob * 0.1f) * 0.1f;
            }
            tmp >>= 4;
            tile_2 <<= 4;
        }
        res = res / num_open;

        if (state.curdepth < CACHE_DEPTH_LIMIT)
        {
            trans_table_entry_t entry = new trans_table_entry_t(state.curdepth, res);
            state.trans_table[board] = entry;
        }

        return res;
    }

    double score_move_node(ref eval_state state, UInt64 board, double cprob)
    {
        double best = 0.0f;

        state.curdepth++;
        for (int move = 0; move < 4; ++move)
        {
            UInt64 newboard = execute_move(board, move);

            state.moves_evaled++;
            if (board != newboard)
            {
                double tmp = score_tilechoose_node(ref state, newboard, cprob);
                if (best < tmp)
                    best = tmp;
            }
            else
            {
                state.nomoves++;
            }
        }
        state.curdepth--;

        return best;
    }

    double score_toplevel_move(UInt64 board, int move)
    {
        double res = 0.0f;
        eval_state state = new eval_state();
        UInt64 newboard = execute_move(board, move);

        state.depth_limit = get_depth_limit(board);

        if (board != newboard)
            res = score_tilechoose_node(ref state, newboard, 1.0f) + 1e-6f;

        Console.WriteLine("Move {0}: result {1}: eval'd {2} moves ({3} no moves, {4} table hits, {5} cache hits, {6} cache size) (maxdepth={7})", move, res,
               state.moves_evaled, state.nomoves, state.tablehits, state.cachehits, state.trans_table.Count, state.maxdepth);

        return res;
    }

    class thrd_context
    {
        public UInt64 board;
        public int move;
        public double res;
        public AutoResetEvent ev = new AutoResetEvent(false);
    }

    void thrd_worker(Object state)
    {
        thrd_context context = (thrd_context)state;
        context.res = score_toplevel_move(context.board, context.move);
        context.ev.Set();
    }

    int find_best_move(UInt64 board)
    {
        double best = 0.0f;
        int bestmove = -1;

        print_board(board);
        Console.WriteLine("Current scores: heur {0}, actual {1}", (UInt32)score_heur_board(board), (UInt32)score_board(board));

        thrd_context[] context = new thrd_context[4];
        for (int move = 0; move < 4; move++)
        {
            context[move] = new thrd_context();
            context[move].board = board;
            context[move].move = move;
            context[move].res = 0.0f;
            ThreadPool.QueueUserWorkItem(thrd_worker, context[move]);
        }
        for (int move = 0; move < 4; move++)
        {
            context[move].ev.WaitOne();
            if (context[move].res > best)
            {
                best = context[move].res;
                bestmove = move;
            }
        }
        Console.WriteLine("Selected bestmove: {0}, result: {1}", bestmove, best);

        return bestmove;
    }

    void play_game(get_move_func_t get_move)
    {
        UInt64 board = initial_board();
        int scorepenalty = 0;
        int last_score = 0, current_score = 0, moveno = 0;

        init_tables();
        while (true)
        {
            int move = 0;
            UInt64 tile = 0;
            UInt64 newboard;

            clear_screen();
            for (move = 0; move < 4; move++)
            {
                if (execute_move(board, move) != board)
                    break;
            }
            if (move == 4)
                break;

            current_score = (int)(score_board(board) - scorepenalty);
            Console.WriteLine("Move #{0}, current score={1}(+{2})", ++moveno, current_score, current_score - last_score);
            last_score = current_score;

            move = get_move(board);
            if (move < 0)
                break;

            newboard = execute_move(board, move);
            if (newboard == board)
            {
                moveno--;
                continue;
            }

            tile = draw_tile();
            if (tile == 2)
                scorepenalty += 4;

            board = insert_tile_rand(newboard, tile);
        }

        print_board(board);
        Console.WriteLine("Game over. Your score is {0}.", current_score);
    }

    static void Main(string[] args)
    {
        Game2048 game_2048 = new Game2048();
        game_2048.play_game(game_2048.find_best_move);
    }
}

File: python/2048-ai.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

import os
import sys
import random
import math

random.seed()

def unif_random(n):
    return random.randrange(n)

def clear_screen():
    if os.name == 'posix':
        _ = os.system('clear')
    else:
        _ = os.system('cls')

ROW_MASK = 0xFFFF
COL_MASK = 0x000F000F000F000F
INT64_MASK = 0xFFFFFFFFFFFFFFFF

UP = 0
DOWN = 1
LEFT = 2
RIGHT = 3

TABLESIZE = 65536
row_left_table = [0] * TABLESIZE
row_right_table = [0] * TABLESIZE
score_table = [0] * TABLESIZE
score_heur_table = [0] * TABLESIZE

SCORE_LOST_PENALTY = 200000.0
SCORE_MONOTONICITY_POWER = 4.0
SCORE_MONOTONICITY_WEIGHT = 47.0
SCORE_SUM_POWER = 3.5
SCORE_SUM_WEIGHT = 11.0
SCORE_MERGES_WEIGHT = 700.0
SCORE_EMPTY_WEIGHT = 270.0
CPROB_THRESH_BASE = 0.0001
CACHE_DEPTH_LIMIT = 15

def unpack_col(row):
    return (row | (row << 12) | (row << 24) | (row << 36)) & COL_MASK

def reverse_row(row):
    return ((row >> 12) | ((row >> 4) & 0x00F0)  | ((row << 4) & 0x0F00) | (row << 12)) & ROW_MASK

def print_board(board):
    sys.stdout.write('-----------------------------%s' % os.linesep)
    for i in range(0, 4):
        for j in range(0, 4):
            power_val = board & 0xf
            if power_val == 0:
                sys.stdout.write('|%6c' % ' ')
            else:
                sys.stdout.write("|%6d" % (1 << power_val))
            board >>= 4
        sys.stdout.write('|%s' % os.linesep)
    sys.stdout.write('-----------------------------%s' % os.linesep)

def transpose(x):
    a1 = x & 0xF0F00F0FF0F00F0F
    a2 = x & 0x0000F0F00000F0F0
    a3 = x & 0x0F0F00000F0F0000
    a = a1 | (a2 << 12) | (a3 >> 12)
    b1 = a & 0xFF00FF0000FF00FF
    b2 = a & 0x00FF00FF00000000
    b3 = a & 0x00000000FF00FF00
    return b1 | (b2 >> 24) | (b3 << 24)

def count_empty(x):
    x |= (x >> 2) & 0x3333333333333333
    x |= (x >> 1)
    x = ~x & 0x1111111111111111
    x += x >> 32
    x += x >> 16
    x += x >> 8
    x += x >> 4
    return x & 0xf

class trans_table_entry_t:
    def __init__(self, depth, heuristic):
        self.depth = depth
        self.heuristic = heuristic

class eval_state:
    def __init__(self):
        self.trans_table = {}
        self.maxdepth = 0
        self.curdepth = 0
        self.nomoves = 0
        self.tablehits = 0
        self.cachehits = 0
        self.moves_evaled = 0
        self.depth_limit = 0

def init_tables():
    for row in range(0, 65536):
        result = 0
        rev_row = 0
        rev_result = 0
        score = 0
        line = [row & 0xf, (row >> 4) & 0xf, (row >> 8) & 0xf, (row >> 12) & 0xf]

        for i in range(0, 4):
            rank = line[i]
            if rank >= 2:
                score += (rank - 1) * (1 << rank)
        score_table[row] = score

        sum_ = 0.0
        empty = 0
        merges = 0
        prev = 0
        counter = 0
        for i in range(0, 4):
            rank = line[i]

            sum_ += math.pow(rank, SCORE_SUM_POWER)
            if rank == 0:
                empty += 1
            else:
                if prev == rank:
                    counter += 1
                elif counter > 0:
                    merges += 1 + counter
                    counter = 0
                prev = rank

        if counter > 0:
            merges += 1 + counter

        monotonicity_left = 0.0
        monotonicity_right = 0.0

        for i in range(1, 4):
            if line[i - 1] > line[i]:
                monotonicity_left += math.pow(line[i - 1], SCORE_MONOTONICITY_POWER) - math.pow(line[i], SCORE_MONOTONICITY_POWER)
            else:
                monotonicity_right += math.pow(line[i], SCORE_MONOTONICITY_POWER) - math.pow(line[i - 1], SCORE_MONOTONICITY_POWER)

        score_heur_table[row] = SCORE_LOST_PENALTY + SCORE_EMPTY_WEIGHT * empty + SCORE_MERGES_WEIGHT * merges - \
            SCORE_MONOTONICITY_WEIGHT * min(monotonicity_left, monotonicity_right) - SCORE_SUM_WEIGHT * sum_

        i = 0
        while i < 3:
            j = i + 1
            while j < 4:
                if line[j] != 0:
                    break
                j += 1
            if j == 4:
                break

            if line[i] == 0:
                line[i] = line[j]
                line[j] = 0
                i -= 1
            elif line[i] == line[j]:
                if line[i] != 0xf:
                    line[i] += 1
                line[j] = 0
            i += 1
        result = line[0] | (line[1] << 4) | (line[2] << 8) | (line[3] << 12)

        rev_row = reverse_row(row)
        rev_result = reverse_row(result)
        row_left_table[row] = row ^ result
        row_right_table[rev_row] = rev_row ^ rev_result

def execute_move(board, move):
    ret = board
    if move == UP:
        t = transpose(board)
        ret ^= unpack_col(row_left_table[t & ROW_MASK])
        ret ^= unpack_col(row_left_table[(t >> 16) & ROW_MASK]) << 4
        ret ^= unpack_col(row_left_table[(t >> 32) & ROW_MASK]) << 8
        ret ^= unpack_col(row_left_table[(t >> 48) & ROW_MASK]) << 12
    elif move == DOWN:
        t = transpose(board)
        ret ^= unpack_col(row_right_table[t & ROW_MASK])
        ret ^= unpack_col(row_right_table[(t >> 16) & ROW_MASK]) << 4
        ret ^= unpack_col(row_right_table[(t >> 32) & ROW_MASK]) << 8
        ret ^= unpack_col(row_right_table[(t >> 48) & ROW_MASK]) << 12
    elif move == LEFT:
        ret ^= row_left_table[board & ROW_MASK]
        ret ^= row_left_table[(board >> 16) & ROW_MASK] << 16
        ret ^= row_left_table[(board >> 32) & ROW_MASK] << 32
        ret ^= row_left_table[(board >> 48) & ROW_MASK] << 48
    elif move == RIGHT:
        ret ^= row_right_table[board & ROW_MASK]
        ret ^= row_right_table[(board >> 16) & ROW_MASK] << 16
        ret ^= row_right_table[(board >> 32) & ROW_MASK] << 32
        ret ^= row_right_table[(board >> 48) & ROW_MASK] << 48
    return ret

def score_helper(board, table):
    return table[board & ROW_MASK] + table[(board >> 16) & ROW_MASK] + \
           table[(board >> 32) & ROW_MASK] + table[(board >> 48) & ROW_MASK]

def score_board(board):
    return score_helper(board, score_table)

def score_heur_board(board):
    return score_helper(board, score_heur_table) + score_helper(transpose(board), score_heur_table)

def draw_tile():
    rd = unif_random(10)
    if rd < 9:
       return 1
    else:
       return 2

def insert_tile_rand(board, tile):
    index = unif_random(count_empty(board))
    tmp = board
    while True:
        while (tmp & 0xf) != 0:
            tmp >>= 4
            tile <<= 4
        if index == 0:
            break
        index -= 1
        tmp >>= 4
        tile <<= 4
    return (board | tile) & INT64_MASK

def initial_board():
    board = draw_tile() << (unif_random(16) << 2)
    return insert_tile_rand(board, draw_tile())

def get_depth_limit(board):
    bitset = 0

    while board != 0:
        bitset |= 1 << (board & 0xf)
        board >>= 4

    if bitset <= 128:
        return 1
    elif bitset <= 512:
        return 2
    return 3

def score_tilechoose_node(state, board, cprob):
    if (cprob < CPROB_THRESH_BASE) or (state.curdepth >= state.depth_limit):
        state.maxdepth = max(state.curdepth, state.maxdepth)
        state.tablehits += 1
        return score_heur_board(board)

    if state.curdepth < CACHE_DEPTH_LIMIT:
        if board in state.trans_table:
            entry = state.trans_table[board]

            if entry.depth <= state.curdepth:
                state.cachehits += 1
                return entry.heuristic

    num_open = count_empty(board)
    cprob /= num_open
    res = 0.0
    tmp = board
    tile_2 = 1
    while tile_2 != 0:
        if (tmp & 0xf) == 0:
            res += score_move_node(state, board | tile_2, cprob * 0.9) * 0.9
            res += score_move_node(state, board | ((tile_2 << 1) & INT64_MASK), cprob * 0.1) * 0.1
        tmp >>= 4
        tile_2 <<= 4
        tile_2 &= INT64_MASK
    res = res / num_open

    if state.curdepth < CACHE_DEPTH_LIMIT:
        state.trans_table[board] = trans_table_entry_t(state.curdepth, res)

    return res

def score_move_node(state, board, cprob):
    best = 0.0
    state.curdepth += 1
    for move in range(0, 4):
        newboard = execute_move(board, move)
        state.moves_evaled += 1
        if board != newboard:
            best = max(best, score_tilechoose_node(state, newboard, cprob))
        else:
            state.nomoves += 1
    state.curdepth -= 1
    return best

def _score_toplevel_move(state, board, move):
    newboard = execute_move(board, move)
    if (board == newboard):
        return 0.0
    return score_tilechoose_node(state, newboard, 1.0) + 0.000001

def score_toplevel_move(board, move):
    res = 0.0
    state = eval_state()
    state.depth_limit = get_depth_limit(board)
    res = _score_toplevel_move(state, board, move)
    sys.stdout.write("Move %d: result %f: eval'd %d moves (%d no moves, %d table hits, %d cache hits, %d cache size) (maxdepth=%d)%s" % (move, res,
           state.moves_evaled, state.nomoves, state.tablehits, state.cachehits, len(state.trans_table), state.maxdepth, os.linesep))
    return res

def find_best_move(board):
    best = 0.0
    bestmove = -1
    print_board(board)
    sys.stdout.write("Current scores: heur %d, actual %d%s" % (score_heur_board(board), score_board(board), os.linesep))
    for move in range(0, 4):
        res = score_toplevel_move(board, move)
        if res > best:
            best = res
            bestmove = move
    sys.stdout.write("Selected bestmove: %d, result: %f%s" % (bestmove, best, os.linesep))
    return bestmove

def play_game(get_move):
    board = initial_board()
    scorepenalty = 0
    last_score = 0
    current_score = 0
    moveno = 0

    init_tables()
    while True:
        clear_screen()
        move = 0
        while move < 4:
            if execute_move(board, move) != board:
                break
            move += 1
        if move == 4:
            break

        current_score = score_board(board) - scorepenalty
        moveno += 1
        sys.stdout.write('Move #%d, current score=%d(+%d)%s' % (moveno, current_score, current_score - last_score, os.linesep))
        last_score = current_score
        sys.stdout.flush()

        move = get_move(board)
        if move < 0:
            break

        newboard = execute_move(board, move)
        if newboard == board:
            moveno -= 1
            continue

        tile = draw_tile()
        if tile == 2:
            scorepenalty += 4

        board = insert_tile_rand(newboard, tile)

    print_board(board)
    sys.stdout.write('Game over. Your score is %d.%s' % (current_score, os.linesep))
    sys.stdout.flush()


if __name__ == "__main__":
    play_game(find_best_move)
